tasks/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+
+                    let body = match StatusV2::from(status) {
+                        StatusV2::Tagged(Tagged { body, .. }) => body,
+                        _ => unreachable!(),
+                    };
+
+                    let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                    let output = Some(task.process_tagged(body));
+
+                    return SchedulerEvent::TaskFinished(TaskToken { handle, output });
+                }
+                ClientFlowEvent::DataReceived { data } => {
+                    if let Some(data) = trickle_down(data, self.active_tasks_mut(), |task, data| {
+                        task.process_data(data)
+                    }) {
+                        return SchedulerEvent::Unsolicited(Response::Data(data));
+                    }
+                }
+                ClientFlowEvent::StatusReceived { status } => match StatusV2::from(status) {
+                    StatusV2::Untagged(body) => {
+                        if let Some(body) =
+                            trickle_down(body, self.active_tasks_mut(), |task, body| {
+                                task.process_untagged(body)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Untagged(body),
+                            )));
+                        }
+                    }
+                    StatusV2::Bye(bye) => {
+                        if let Some(bye) =
+                            trickle_down(bye, self.active_tasks_mut(), |task, bye| {
+                                task.process_bye(bye)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Bye(bye),
+                            )));
+                        }
+                    }
+                    StatusV2::Tagged(Tagged { tag, body }) => {
+                        let handle = {
+                            let (handle, _) = self
+                                .active_tasks
+                                .iter()
+                                .find(|(_, (tag_, _))| tag == *tag_)
+                                .unwrap();
+
+                            *handle
+                        };
+
+                        let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                        let output = Some(task.process_tagged(body));
+
+                        return SchedulerEvent::TaskFinished(TaskToken { handle, output });
+                    }
+                },
+            }
+        }
+    }
+
+    fn active_tasks_mut(&mut self) -> impl Iterator<Item = &mut Box<dyn TaskAny>> {
+        self.active_tasks.values_mut().map(|(_, task)| task)
+    }
+}
+
+#[derive(Debug)]
+pub enum SchedulerEvent {
+    TaskFinished(TaskToken),
+    Shutdown,
```

Comment:
What's this?

Reply:
The thing I could use instead of `FatalUnknownTagStatusFromCthulhu`? :-)

Reply:
I did something different. The reason for the `Shutdown` variant was that an IMAP server can send a `bye` at any time. A client receving an unsolicited bye is then expected to read as many responses as it can (until the server closes the connection). This is an expected outcome and not an error. Without `Shutdown` we basically assume a connection lives forever. Still: Let's postpone this discussion. I'll remove `Shutdown` for now.

Reply:
Forget what I said... delted. Shutdown is, in fact, `Unsolicited(Bye)?. There should be no need to signal it separately.